### ISSUE TITLE
[Static extensions] Should we allow static extensions to be accessed on any named type?

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/language#4055]
The current proposal (#3835) for static extensions only allows static members and constructors to be accessed only on "class-like" types (that is classes, mixin classes, mixins, enums, and extension types).  In principle, we could relax this restriction, either in the case that the `on` type is a typedef (see #4052) or in the case that the receiver is a typedef name.  For example, we could allow:

```dart
typedef Pair<T> = (T, T);

extension E<T> on Pair<T> {
   factory Pair.twice(T x) => (x, x);
}
void test() {
   var (x, y) = Pair.twice(3);
   assert(3 == y);
   assert(3 == x);
}
```

If so, do we attach the extension to the name `Pair` or to the underlying type?  That is, does the following work?

```dart
typedef Pair<T> = (T, T);

extension E<T> on (T, T) {
   factory Pair.twice(T x) => (x, x);
}
void test() {
   var (x, y) = Pair.twice(3);
   assert(3 == y);
   assert(3 == x);
}
```

cc @dart-lang/language-team

**Comment by lrhn** [comment]
No. I'd only allow adding extension static members to existing static namespaces. 

If we say "any type can have static members", we'll have to explain why `List<int>` and `List<String>` do not have different static namespaces (or even worse, but `List<int> Function()` and `List<String> Function ()` do. (Or say how to collapse those function type namespaces into one.)

The static namespaces are introduced by `class`, `mixin`, `enum`, `extension type` and `extension` declarations. 

Some type aliases have a shape which we say denotes a static namespace, and we let them also alias that existing static namespace. (Which cannot be an extension namespace, since those do not introduce a type.)

With this feature, some extensions will also act as extensions of the static namespaces of their `on` type, if the `on` type clause has a shape that denotes a static namespace. 
When we would otherwise do a static namespace lookup on that namespace, and we find no member with the needed name, we check if any extension extends that static namespace with something of the same base name, and if a single one is applicable, the static reference denotes that static member definition.

(And if we did, we would not attach anything to type alias names.)
